File: web/src/api/OrganizationApi.ts
import Api, { apiState } from 'api/Api';
import { selector } from 'recoil';
import OrganizationRep from 'rep/OrganizationRep';

class OrganizationApi {
  private readonly api: Api;

  constructor(api: Api) {
    this.api = api;
  }

  async get(organizationId: string): Promise<OrganizationRep | null> {
    return await this.api.request<OrganizationRep | null>({
      method: 'GET',
      path: `/organizations/${organizationId}`,
    });
  }

  async listAll(): Promise<OrganizationRep[]> {
    return await this.api.request<OrganizationRep[]>({
      method: 'GET',
      path: '/organizations',
    });
  }

  async create(creator: OrganizationRep.Creator): Promise<OrganizationRep> {
    return await this.api.request<OrganizationRep, OrganizationRep.Creator>({
      method: 'POST',
      path: '/organizations',
      body: creator,
    });
  }

  async update(organizationId: string, updater: OrganizationRep.Updater): Promise<OrganizationRep> {
    return await this.api.request<OrganizationRep, OrganizationRep.Updater>({
      method: 'PATCH',
      path: `/organizations/${organizationId}`,
      body: updater,
    });
  }

  async delete(organizationId: string): Promise<OrganizationRep> {
    return await this.api.request<OrganizationRep>({
      method: 'DELETE',
      path: `/organizations/${organizationId}`,
    });
  }
}

export const organizationApiState = selector<OrganizationApi>({
  key: 'api/organization',
  get: ({ get }) => {
    const api = get(apiState({ authenticated: true }));
    return new OrganizationApi(api);
  },
});

File: feature/organization/rest-interface/src/main/kotlin/limber/api/organization/OrganizationApi.kt
package limber.api.organization

import io.ktor.http.HttpMethod
import jakarta.validation.Valid
import limber.feature.rest.QueryParam
import limber.feature.rest.RestEndpoint
import limber.rep.organization.OrganizationRep
import limber.validation.SearchValidator

public object OrganizationApi {
  public data class Get(
    val organizationId: String,
  ) : RestEndpoint<Nothing>() {
    override val method: HttpMethod = HttpMethod.Get
    override val path: String = "/organizations/$organizationId"
  }

  public data object ListAll : RestEndpoint<Nothing>() {
    override val method: HttpMethod = HttpMethod.Get
    override val path: String = "/organizations"
  }

  public data class Search(
    @SearchValidator val search: String,
  ) : RestEndpoint<Nothing>() {
    override val method: HttpMethod = HttpMethod.Get
    override val path: String = "/organizations"
    override val qp: List<QueryParam> = listOf(::search)
  }

  public data class Create(
    @Valid override val body: OrganizationRep.Creator?,
  ) : RestEndpoint<OrganizationRep.Creator>() {
    override val method: HttpMethod = HttpMethod.Post
    override val path: String = "/organizations"
  }

  public data class Update(
    val organizationId: String,
    @Valid override val body: OrganizationRep.Update?,
  ) : RestEndpoint<OrganizationRep.Update>() {
    override val method: HttpMethod = HttpMethod.Patch
    override val path: String = "/organizations/$organizationId"
  }
}
